src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -1216,4 +1250,31 @@ public long countSubmissions(Long exerciseId) {
     public long countSubmissionsToAssessByCourseId(Long courseId) {
         return programmingExerciseRepository.countByCourseIdSubmittedBeforeDueDate(courseId);
     }
+
+    private Map<Path, ScheduledFuture> futures = new HashMap<>();
+
+    private ScheduledExecutorService executor = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
```

Comment:
I would suggest to also use a native Spring task scheduler here. We can either use the existing scheduler (currently only has `poolSize=1`, although this should be fine since we don't have a high frequency of tasks. We could also just increase the poolSize), or have a dedicated config for this scenario. Either way, imo injecting a scheduler bean is more consistent than using the native Java pools.

Reply:
Good idea for a subsequent PR